### PR TITLE
Validation improvements

### LINKED ISSUE #191: Making static members that hide a superclass member is impossible
Java allows hiding a superclass method `A.doSomething()` with a subclass method `B.doSomething()`, but Enigma doesn't allow giving static methods names that do that.

See also https://github.com/FabricMC/yarn/pull/1040#discussion_r364046631 and https://github.com/FabricMC/yarn/pull/858#discussion_r324467245.

### LINKED ISSUE #246: Opening the rename box for MinecraftVersion and pressing enter without changing the name causes "Name is not unique" error
Screenshot is from my tabs PR branch (#238), but the same issue occurs in 2b906a2, just the error message is different ("Name is not unique!"). Issue occurs with https://github.com/FabricMC/yarn/commit/bfbea641fcf1b21ea1b346b87d12525daca3c476

This does not occur with any other class as far as I'm aware.

image

### PR DESCRIPTION
**Pull request by thiakil**
Made some improvements to MappingValidator to take access levels into account when checking name uniqueness.
Should hopefully fix #191 and fix #246 

Also added a Test to ensure this doesn't get broken, etc.

### CODE REVIEW

File: enigma/src/main/java/cuchaz/enigma/translation/mapping/MappingValidator.java [hunk 1]
```diff
@@ -35,45 +39,105 @@ public void validateRename(ValidationContext vc, Entry<?> entry, String name) {
...
+		InheritanceIndex inheritanceIndex = index.getInheritanceIndex();
 
-			if (!isUnique(translatedEntry, translatedSiblings, name)) {
-				Entry<?> parent = translatedEntry.getParent();
+		//Step 1, check it's unique within its own siblings
+		Collection<Entry<?>> directTranslatedSiblings = obfToDeobf.getSiblings(entry).stream()
+				.map(deobfuscator::translate)
+				.collect(Collectors.toList());
+		for (Entry<?> sibling : directTranslatedSiblings) {
+			if (entry.canConflictWith(sibling) && sibling.getName().equals(name) && isSynthetic(entry) == isSynthetic(sibling)) {
```
Comment (thiakil) [inline enigma/src/main/java/cuchaz/enigma/translation/mapping/MappingValidator.java:49]: Not entirely sure on the synthetic check - I wasn't able to reproduce it with the unit test, only with an SRG named MC jar (and an unrelated change I removed from this PR as it was probably broken).

The idea is that if one entry is a synthetic/bridge and the other is not, that we can safely accept the rename (as it's probably only the return type that differs)

Comment (2xsaiko) [reply enigma/src/main/java/cuchaz/enigma/translation/mapping/MappingValidator.java:49]: I think that should be `[..] && !isSynthetic(entry) && !isSynthetic(sibling)`, since if any of the methods is synthetic, they don't conflict since at least one of them is invisible from source code

**Comment by 2xsaiko** [comment]
Should I merge this or do you want to apply my suggestion, @thiakil?

**Comment by thiakil** [comment]
Probably quicker if you merge and then apply it xD

**Comment by 2xsaiko** [comment]
Alright haha

**Comment by 2xsaiko** [comment]
Thanks!

**Comment by 2xsaiko** [comment]
Hmm... This broke the build: https://jenkins.modmuss50.me/blue/organizations/jenkins/FabricMC%2FEnigma/detail/master/222/pipeline
Any idea?

**Comment by thiakil** [comment]
Not sure, I don't think I removed anything from tests

**Comment by 2xsaiko** [comment]
It has to do with build/test-obf not existing (see [this][1]). I tried making a task that creates that directory and making the TestObf task depend on that, but for some reason it didn't create the directory (it did get executed though)... so now you get 0b31b58b7950cfef3c957190ae2038614b5bf79d. It's a bad fix, but I don't want to try to work around gradle's bullshit for half an hour. If you have a better idea, please tell me lol

[1]: https://github.com/FabricMC/Enigma/pull/299/files#diff-d7098b3be8c22cc8f2b2242a587851e5R56

**Comment by thiakil** [comment]
Hmm I'm sure I used that folder because it was used elsewhere in the tests

**Comment by 2xsaiko** [comment]
Yeah, it does work the way you had it if you run `build` only and not `clean build` which is what CI does. No idea why.

**Comment by 2xsaiko** [comment]
I reverted this for now since it let you name two fields in the same class the same.

**Comment by thiakil** [comment]
What two fields? This is allowed in certain circumstances - see the test input

**Comment by 2xsaiko** [comment]
Looks like any. Methods too:
image

**Comment by 2xsaiko** [comment]
Ahh, I know why.

    if (entry.canConflictWith(sibling) && sibling.getName().equals(name) && isSynthetic(entry) == isSynthetic(sibling)) {

canConflictWith checks if the containing class is the same. But in this case, sibling is `net/minecraft/MinecraftVersion.x()I` (deobfuscated) and entry is `net/minecraft/class_3797.getPackVersion()I` (obfuscated, the getPackVersion is the obf name)

**Comment by 2xsaiko** [comment]
Also `isAcceptableOverride` should probably not work the way it does. Currently, you can name a method in a class the same as a different method in a superclass, and have them not conflict, which should obviously not happen.

I think I fixed the first issue, not sure if it's the correct way to do so though:

```patch
From ec9f27b5b635b51b8db773b1b0dc4e22624d01c1 Mon Sep 17 00:00:00 2001
From: 2xsaiko <git@dblsaiko.net>
Date: Sat, 3 Oct 2020 10:58:41 +0200
Subject: [PATCH] Try fixing mapping validation

---
 .../enigma/translation/mapping/MappingValidator.java  | 11 +++++------
 1 file changed, 5 insertions(+), 6 deletions(-)

diff --git a/enigma/src/main/java/cuchaz/enigma/translation/mapping/MappingValidator.java b/enigma/src/main/java/cuchaz/enigma/translation/mapping/MappingValidator.java
index 134c05c..1757b12 100644
--- a/enigma/src/main/java/cuchaz/enigma/translation/mapping/MappingValidator.java
+++ b/enigma/src/main/java/cuchaz/enigma/translation/mapping/MappingValidator.java
@@ -42,13 +42,12 @@ public class MappingValidator {
 		InheritanceIndex inheritanceIndex = index.getInheritanceIndex();
 
 		//Step 1, check it's unique within its own siblings
-		Collection<Entry<?>> directTranslatedSiblings = obfToDeobf.getSiblings(entry).stream()
-				.map(deobfuscator::translate)
-				.collect(Collectors.toList());
-		for (Entry<?> sibling : directTranslatedSiblings) {
-			if (entry.canConflictWith(sibling) && sibling.getName().equals(name) && isSynthetic(entry) == isSynthetic(sibling)) {
+		Collection<Entry<?>> directSiblings = obfToDeobf.getSiblings(entry);
+		for (Entry<?> sibling : directSiblings) {
+			Entry<?> translatedSibling = deobfuscator.translate(sibling);
+			if (entry.canConflictWith(sibling) && translatedSibling.getName().equals(name) && isSynthetic(entry) == isSynthetic(sibling)) {
 				// allow clash if one is synthetic and the other is not
-				Entry<?> parent = entry.getParent();
+				Entry<?> parent = translatedSibling.getParent();
 				if (parent != null) {
 					vc.raise(Message.NONUNIQUE_NAME_CLASS, name, parent);
 				} else {
-- 
2.28.0

```

**Comment by thiakil** [comment]
Hmm am on mobile right now, but in any case I guess we need some negative cases in the test suite to ensure such invalid renames are correctly prevented
